src/linux/cgroups2.cpp

```diff
@@ -295,4 +296,152 @@ Try<set<string>> enabled(const string& cgroup)
+    // 3. [ Minor Version Check ] ⤵
+    // 4. [ Major Version Check ] ⤵
+    // 5. [ Allow/Deny Access   ]  ↓
+    //                             ↓
+    // 6. [ NEXT BLOCK          ] ←←
+    //
+    // Either:
+    // 1. The device access is matched by (1,2,3,4) and the Allow/Deny access
+    //    block (5) is executed.
+    // 2. One of (1,2,3,4) does not match the requested access and we skip
+    //    to the next block (6).
+
+    // We only check attributes that are not 100% permissive:
+    // - Type   != all
+    // - Access != rnw
+    // - Minor  != None (AKA "any")
+    // - Major  != None (AKA "any")
+    Entry::Selector selector = entry.selector;
+    Entry::Access access = entry.access;
+
+    bool check_type = selector.type != Entry::Selector::Type::ALL;
+    bool check_access = !access.mknod || !access.read || !access.write;
+    bool check_major = selector.major.isSome();
+    bool check_minor = selector.minor.isSome();
+
+    // Number of instructions to the [NEXT BLOCK]. This is used if a check
+    // fails (meaning this whitelist entry does not apply) and we want to
+    // skip the subsequent checks.
+    short jmp_size = 1 + (check_access ? 3 : 0) + (check_type ? 1 : 0) +
+                     (check_major ? 1 : 0) + (check_minor ? 1 : 0);
+
+    // Check type (r2) against entry.
+    if (check_type) {
+      int bpf_type = selector.type == Entry::Selector::Type::BLOCK
+                       ? BPF_DEVCG_DEV_BLOCK
+                       : BPF_DEVCG_DEV_CHAR;
+
+      program.append({ BPF_JMP_IMM(BPF_JNE, BPF_REG_2, bpf_type, jmp_size) });
+      --jmp_size;
+    }
+    // Check access (r3) against entry.
+    if (check_access) {
+      int bpf_access = 0;
+      bpf_access |= access.read ? BPF_DEVCG_ACC_READ : 0;
+      bpf_access |= access.write ? BPF_DEVCG_ACC_WRITE : 0;
+      bpf_access |= access.mknod ? BPF_DEVCG_ACC_MKNOD : 0;
+
+     program.append({
+          BPF_MOV32_REG (BPF_REG_1, BPF_REG_3),
+          BPF_ALU32_IMM (BPF_AND, BPF_REG_1, bpf_access),
+          BPF_JMP_REG (BPF_JNE, BPF_REG_1, BPF_REG_3, (short) (jmp_size - 2))});
+      jmp_size -= 3;
+    }
+    // Check major version (r4) against entry.
+    if (check_major) {
+      program.append({
+          BPF_JMP_IMM(BPF_JNE, BPF_REG_4, (int) selector.major.get(), jmp_size)});
+      --jmp_size;
+    }
+    // Check minor version (r5) against entry.
+    if (check_minor) {
+      program.append({
+          BPF_JMP_IMM(BPF_JNE, BPF_REG_5, (int) selector.minor.get(), jmp_size)});
+      --jmp_size;
+    }
+
+    // Allow/Deny access block.
+    program.append({
+        BPF_MOV64_IMM(BPF_REG_0, allow ? ALLOW_ACCESS : DENY_ACCESS),
+        BPF_EXIT_INSN()});
+
+    return Nothing();
+  }
+
+  ebpf::Program build()
+  {
+    // Exit instructions.
+    // If no entry granted access, then deny the access.
+    program.append({
+        BPF_MOV64_IMM (BPF_REG_0, DENY_ACCESS),
+        BPF_EXIT_INSN()});
+
+    return program;
```

Comment:
are we expecting our programs to reach these instructions..? if so, when?

Reply:
If no devices are added to a program. If we _don't_ have these instructions and you try and load the program without any device entries into the kernel, it will cause a validation error. It's protective.

Reply:
ah ok, in that case, can we generate this only when needed, for clarity?

Reply:
Actually, reflecting, we reach this point just as described in the comment:
```
// If no entry granted access, then deny the access.
```